src/heimdall/worker.cc

```diff
@@ -0,0 +1,903 @@
+
+    for (const auto& ll : shape) {
+      // Convert point to Web Mercator
+      double merc_x = lon_to_merc_x(ll.lng());
+      double merc_y = lat_to_merc_y(ll.lat());
+
+      // Normalize to 0-1 within tile's mercator bounds
+      double norm_x = (merc_x - projection.tile_merc_minx) / projection.tile_merc_width;
+      double norm_y = (projection.tile_merc_maxy - merc_y) /
+                      projection.tile_merc_height; // Y is inverted in tiles
+
+      // Scale to tile extent and convert to tile coordinates
+      double tile_x = norm_x * projection.tile_extent;
+      double tile_y = norm_y * projection.tile_extent;
+
+      boost::geometry::append(unclipped_line, point_t(tile_x, tile_y));
+    }
+
+    // Clip the line to the tile boundaries using Boost.Geometry
+    // This properly handles lines that cross tile boundaries
+    multi_linestring_t clipped_lines;
+    boost::geometry::intersection(clip_box, unclipped_line, clipped_lines);
+
+    // Skip if no clipped segments (edge is completely outside tile)
+    if (clipped_lines.empty()) {
+      continue;
+    }
+
+    // Process each clipped line segment (there may be multiple if line crosses tile multiple
+    // times)
+    for (const auto& clipped_line : clipped_lines) {
+      // Skip degenerate lines (less than 2 points)
+      if (clipped_line.size() < 2) {
+        continue;
+      }
+
+      // Convert to vtzero points, removing consecutive duplicates
+      std::vector<vtzero::point> tile_coords;
+      tile_coords.reserve(clipped_line.size());
+
+      int32_t last_x = INT32_MIN;
+      int32_t last_y = INT32_MIN;
+
+      for (const auto& pt : clipped_line) {
+        int32_t x = static_cast<int32_t>(std::round(pt.x()));
+        int32_t y = static_cast<int32_t>(std::round(pt.y()));
+
+        // Skip consecutive duplicate points (can happen after rounding)
+        if (x == last_x && y == last_y) {
+          continue;
+        }
+
+        tile_coords.emplace_back(x, y);
+        last_x = x;
+        last_y = y;
+      }
+
+      // Must have at least 2 unique points to create a valid linestring
+      if (tile_coords.size() < 2) {
+        continue;
+      }
+
+      // Check for opposing edge
+      baldr::graph_tile_ptr opp_tile = edge_tile;
+      const DirectedEdge* opp_edge = nullptr;
+      GraphId opp_edge_id = reader_->GetOpposingEdgeId(edge_id, opp_edge, opp_tile);
+
+      // Add feature using the builder
+
+      const volatile baldr::TrafficSpeed* forward_traffic =
+          edge ? &edge_tile->trafficspeed(edge) : nullptr;
+      const volatile baldr::TrafficSpeed* reverse_traffic =
+          opp_edge ? &opp_tile->trafficspeed(opp_edge) : nullptr;
+
+      edges_builder.add_feature(tile_coords, edge_id, edge, opp_edge_id, opp_edge, forward_traffic,
+                                reverse_traffic, edge_info);
+
+      // Collect the end node of this edge for the nodes layer
+      unique_nodes.insert(edge->endnode());
```

Comment:
are we sure we'll actually end up with all nodes here? I mean the list of edges isn't coming from a traversal, it's just a random collection from the bins. so this could end up trying to insert the same node twice (won't bcs it's a set) and the other node might be missing? or maybe we do get all bidirectional ways twice from this range query and it's currently duplicated? I'm a bit confused, I'm sure that would have been obvious when testing this..